### PR TITLE
Robustness

### PR DESCRIPTION
**Pull request by holisticode**
This PR adds the parameter `Robust` to the `Healthy` structure.

`Robust` denotes 

`whether we are connected to a minimum of peers in all the bins we have known peers in`.

It is meant as a discussion PR until we reach consensus on how this parameter will be used.

For reviewers other than @nolash :
**You are welcome to leave comments but this is not yet ready for formal review.**

### CODE REVIEW

File: swarm/network/kademlia_test.go [hunk 1]
```diff
@@ -241,16 +238,152 @@ func (tk *testKademlia) checkHealth(expectHealthy bool, expectSaturation bool) {
...
+
+	// add second connected neighbor
+	// saturated false
+	addr = pot.RandomAddressAt(baseAddress, 4)
+	peer = newTestDiscoveryPeer(addr, k)
+	k.On(peer)
+	isSaturated(t, k, false)
+
+	// connect peer in zero-bin
+	// saturated true
```
Comment (holisticode) [inline swarm/network/kademlia_test.go]: @nolash from here on **all** saturated tests were failing, and I had to **invert** their expected result in order for them to pass (`true` -> `false`,    `false`-> `true`)

In fact based on the `saturated` criteria, which is `k.saturation < depth`, the results we get are correct.
But of course we need to revisit if:
 * It is indeed what we would expect from the given kademlia
 * If the test is actually accurate enough to test for saturation (maybe we need more/other cases?)

Or maybe the whole `IsSaturated` test is **outside the scope of this PR** (which would be about robustness)?

File: swarm/network/kademlia.go [hunk 2]
```diff
@@ -750,18 +754,49 @@ func (k *Kademlia) connectedNeighbours(peers [][]byte) (got bool, n int, missing
...
+	})
+	k.eachConn(nil, 255, func(_ *Peer, po int) bool {
+		pc[po]++
+		return true
+	})
+
+	for po, v := range pk {
+		if pc[po] == v {
+			continue
+		} else if po >= depth && pc[po] != pk[po] {
```
Comment (nolash) [inline swarm/network/kademlia.go]: I realize perhaps this is redundant:

`pc[po] != pk[po]`

This is (should be) the same as `knowNeighbours && connectedNeighbours`

File: swarm/network/kademlia_test.go [hunk 3]
```diff
@@ -241,16 +238,152 @@ func (tk *testKademlia) checkHealth(expectHealthy bool, expectSaturation bool) {
 	pp := NewPeerPotMap(tk.NeighbourhoodSize, addrs)
 	healthParams := tk.Healthy(pp[kid])
 
-	// definition of health, all conditions but be true:
-	// - we at least know one peer
-	// - we know all neighbors
-	// - we are connected to all known neighbors
```
Comment (nolash) [inline swarm/network/kademlia_test.go:247]: Why is the comment gone?

File: swarm/network/kademlia_test.go (hunk 1 above)
Comment (nolash) [reply swarm/network/kademlia_test.go]: If they don't work please remove them, and let @zelig write those tests later.

**Review by nolash — CHANGES_REQUESTED** [review]

File: swarm/network/kademlia_test.go (hunk 1 above)
Comment (holisticode) [reply swarm/network/kademlia_test.go]: Removed

File: swarm/network/kademlia.go (hunk 2 above)
Comment (holisticode) [reply swarm/network/kademlia.go]: I am not sure I understand this - in any case, IS it redundant? Or only perhaps?

File: swarm/network/kademlia_test.go (hunk 3 above)
Comment (holisticode) [reply swarm/network/kademlia_test.go:247]: Moved to the `IsHealthyStrict()` function

File: swarm/network/simulations/discovery/discovery_test.go [hunk 4]
```diff
@@ -267,11 +267,11 @@ func discoverySimulation(nodes, conns int, adapter adapters.NodeAdapter) (*simul
 		}
 
 		healthy := &network.Health{}
-		if err := client.Call(&healthy, "hive_healthy", ppmap[common.Bytes2Hex(id.Bytes())]); err != nil {
+		if err := client.Call(&healthy, "hive_getHealthInfo", ppmap[common.Bytes2Hex(id.Bytes())]); err != nil {
 			return false, fmt.Errorf("error getting node health: %s", err)
 		}
-		log.Debug(fmt.Sprintf("node %4s healthy: connected nearest neighbours: %v, know nearest neighbours: %v,\n\n%v", id, healthy.ConnectNN, healthy.KnowNN, healthy.Hive))
-		return healthy.KnowNN && healthy.ConnectNN, nil
+		log.Debug(fmt.Sprintf("node %4s healthy: connected nearest neighbours: %v, know nearest neighbours: %v, is robuts: %v, \n\n%v", id, healthy.ConnectNN, healthy.KnowNN, healthy.Robust, healthy.Hive))
```
Comment (nolash) [inline swarm/network/simulations/discovery/discovery_test.go]: robuts

File: swarm/network/kademlia.go (hunk 2 above)
Comment (nolash) [reply swarm/network/kademlia.go]: The desciption of this method says:

>  getWeakBins checks whether the node is connected to a health minimum of peers it knows about in bins that are shallower than depth

`po >= depth` is depth or deeper.

So I would say it's redundant.

Comment (holisticode) [reply swarm/network/kademlia.go]: ok removed

**Review by nolash — APPROVED** [review]
For my part this is now ok. But I am not at all sure whether the idea I have of "healthiness" is the right one anymore. So I defer judgement. My advice is to have two more approvals of this besides mine from someone who understands the terms that are in use.

**Comment by zelig** [comment]
I had a discussion with @holisticode about this. IMHO healthy is primarily used as a stop condition in simulation tests
Since the current version only checks nearest neighbours, saturation of bins shallower than depth is not checked. 
With another implementation of hive discovery, unhealthy kademlias could pass as healthy. 
But saturation check cannot be absolute since we can have situations that there is simply less than MinBinSize peers for a node in a bin. Therefore the peer pot has to be enriched with the counts of potential peers for each bin (upto depth at least) for each node.
These cardinalities can be passed to the health function via a slice of integers which are compared to known and connected peers.

**Comment by zelig** [comment]
obsoleted by #1195

**Pull request closed without merging**